---
title: "Project"
author: "Trude Almestrand"
date: "8 11 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Project

### Abstract
### Introduction
### Methods
Supervised learning: 
(a) PCA/hierarchal clustering/heatmap to get an initial overview of the data, 
(b) differential expression analysis to reduce the dimensionality, 
(c) machine learning to predict outcome, 
(d) extract some biological insight from the machine learning model


### Results
### Discussion
### References

http://www.sthda.com/english/articles/37-model-selection-essentials-in-r/152-principal-component-and-partial-least-squares-regression-essentials/ 
https://machinelearningmastery.com/feature-selection-with-real-and-categorical-data/ 
https://www.r-bloggers.com/2014/05/evaluating-model-performance-a-practical-example-of-the-effects-of-overfitting-and-data-size-on-prediction/ 

## Extra

Including code
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("TCGAbiolinks")

library(TCGAbiolinks)

# View all projects
# View(TCGAbiolinks:::getGDCprojects())

expr <- data.frame()
samples <- c()
classes <- c()

# Download RNA-Seq data for two selected projects
tissue <- c("Skin", "Testis", "Ovary")
proj   <- c("TCGA-SKCM", "TCGA-TGCT", "TCGA-OV")
for (i in 1:length(proj)) {

  query <- GDCquery(project = proj[i],
                    data.category = "Gene expression",
                    data.type = "Gene expression quantification",
                    platform = "Illumina HiSeq", 
                    file.type = "results",
                    experimental.strategy = "RNA-Seq",
                    legacy = TRUE)
  GDCdownload(query, method = "api", files.per.chunk = 10)
  
  data <- GDCprepare(query, )
  tmp.expr <- assay(data)
  tmp.samples <- paste(tissue[i], colData(data)$primary_diagnosis, sep="-")
  
  # Sample must have 20M reads
  idx <- colSums(tmp.expr) > 20E6
  tmp.expr <- tmp.expr[,idx]
  tmp.samples <- tmp.samples[idx]
  
  # Randomly select 100 samples
  idx <- sample(1:ncol(tmp.expr), 100)
  tmp.expr <- tmp.expr[,idx]
  tmp.samples <- tmp.samples[idx]
  
  if (nrow(expr) == 0) {
    expr <- tmp.expr
  } else {
    expr <- cbind(expr, tmp.expr)
  }
  samples <- c(samples, tmp.samples)
  classes <- c(classes, rep(tissue[i], length(tmp.samples)))
}
remove(tmp.expr, tmp.samples)
classes <- factor(classes)

table(classes)
dim(expr)

# Filtering
# Genes must have 1024 mapped reads in at least 10 samples
expr <- expr[rowSums(expr > 2^10) > 10,]
dim(expr)

save(expr, classes, file = "Project_TCGA.RData")
```

## PCA
```{r}
library(beadarray)
library(DESeq2)

expr <- round(expr)

SampleTable <- data.frame(classes)
dds <- DESeqDataSetFromMatrix(expr, colData = SampleTable, design = ~classes)

# Sets sample name according to where it was retrieved
colnames(expr) <- classes
#Performs normalization
expr.norm <- varianceStabilizingTransformation(dds)

# Runs PCA
expr.pca <- prcomp(t(assay(expr.norm)))

#Plots PCA
plot(expr.pca$x[, 1], expr.pca$x[, 2], xlab = "PC1", ylab = "PC2", pch = 16, col = classes)
legend("topright", legend = levels(classes), pch = 16, col = 1:5)
```

